language/bytecode-verifier/src/reference_safety/mod.rs

```diff
@@ -324,6 +324,62 @@ fn execute_inner(
             let struct_def = verifier.resolver.struct_def_at(struct_inst.def)?;
             unpack(verifier, struct_def)
         }
+
+        Bytecode::VecEmpty(idx) => {
+            // since reference safety is scheduled after type safety, we know that the element type
+            // must be valid, i.e., the unwrap is safe here.
+            let element_type = verifier
+                .resolver
+                .signature_at(*idx)
+                .0
+                .last()
+                .unwrap()
+                .clone();
+            verifier
+                .stack
+                .push(state.value_for(&SignatureToken::Vector(Box::new(element_type))));
+        }
+
+        Bytecode::VecLen => {
+            let vec_ref = verifier.stack.pop().unwrap();
+            state.vector_op(offset, vec_ref, false)?;
+            verifier.stack.push(state.value_for(&SignatureToken::U64));
+        }
+
+        Bytecode::VecImmBorrow => {
+            checked_verify!(verifier.stack.pop().unwrap().is_value());
+            let vec_ref = verifier.stack.pop().unwrap();
+            state.vector_element_borrow(offset, vec_ref, false)?;
```

Comment:
We might want to relax this should be relaxed to match `ImmBorrowField` in that it should also work over mutable vector references.

That being said... the relaxation doesn't do anything here. There are times where ImmBorrowField is valid but freeze is not, hence that relaxed rule. Thoughts, @shazqadeer @sblackshear ?

Reply:
I'm not sure I understand "the relaxation doesn't do anything here"...can you elaborate?

Reply:
`ImmBorrowField` can work over mutable or immutable references. This is due to adding expressibility where a given field can be borrowed immutably, but the parent ref cannot be frozen. 

As written here, `VecImmBorrow` can work only over immutable references

Reply:
Updated, now both `VecImmBorrow` and `VecLen` accept a mutable reference of a vector type as well.